### PR TITLE
Remove IOdspSnapshot, IOdspSnapshotCommit, IOdspSnapshotBlob from odsp driver.

### PR DESCRIPTION
**Pull request by jatgarg**
Refer: https://github.com/microsoft/FluidFramework/issues/4589 (Issue)
Refer: https://github.com/microsoft/FluidFramework/pull/6560 (Other PR which reads newer snapshot format)

Now storage manager just see the ISnapshotTree and ArrayBuffer for blobs as these are used in runtime too. The old IOdspSnapshot will just be hidden to the fetcher as it is the only thing which knows what it has fetched. This will help to move to newer version of snapshot format.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 1]
```diff
@@ -41,8 +42,8 @@ export async function fetchSnapshot(
     versionId: string,
     fetchFullSnapshot: boolean,
     logger: ITelemetryLogger,
-    snapshotDownloader: (url: string, fetchOptions: {[index: string]: any}) => Promise<IOdspResponse<IOdspSnapshot>>,
-): Promise<IOdspResponse<IOdspSnapshot>> {
+    snapshotDownloader: <T>(url: string, fetchOptions: {[index: string]: any}) => Promise<IOdspResponse<T>>,
+): Promise<ISnapshotCacheValueV2> {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:46]: I'd not tie the response of fetching to caching. You can define almost identical interface but without version and use it for most of the communication (including here), but only add version right before adding it into the cache. Same with reading - version is only required by version layer, beyond that there is no need to leak that detail.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 2]
```diff
@@ -143,9 +145,9 @@ async function fetchLatestSnapshotCore(
     storageTokenFetcher: (options: TokenFetchOptions, name: string) => Promise<string | null>,
     snapshotOptions: ISnapshotOptions | undefined,
     logger: ITelemetryLogger,
-    snapshotDownloader: (url: string, fetchOptions: {[index: string]: any}) => Promise<IOdspResponse<IOdspSnapshot>>,
+    snapshotDownloader: <T>(url: string, fetchOptions: {[index: string]: any}) => Promise<IOdspResponse<T>>,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: I'm a bit confused about usage of T here.
I'd expect compiler to require it to be generics arg for a function (fetchLatestSnapshotCore)
But also usage below assumes only IOdspSnapshot, so should T be removed (at a generic argument) and replaces (return value) with IOdspSnapshot?

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 3]
```diff
@@ -205,7 +207,7 @@ async function fetchLatestSnapshotCore(
             },
             async (event) => {
                 const startTime = performance.now();
-                const response: IOdspResponse<IOdspSnapshot> = await snapshotDownloader(
+                const response = await snapshotDownloader<IOdspSnapshot>(
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:210]: Would it be better to push requirement to convert into snapshotDownloader, such that this function and others are abstracted away from wire format? That way plugging in new format would be much simpler (as of now, all of these functionality would need to be duplicated).
Basically there are three layers:
1. Generic fetch functionality - does not know what it fetches
2. Middle layer, that works with specific wire format and does conversion to "common" format. This layer is ideally just one function.
3. All the rest of the code (majority of it) that does not know anything about wire format.

This function I believe belongs to # 3.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: Here snapshotDownloader function is a generic which can return different types based on T like IOdspSnapshot or the newer format. It does not need to respond with IOdspsnapshot as the new format is coming. So the caller will know what type of result will it be.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: Let's assume that we now have two formats. How this function (fetchSnapshotWithRedeem) knows what format it gets from downloader? And why responsibility of conversion is not pushed to downloader (or even one layer below)?

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:210]: Yes in this: snapshotDownloader belongs to #1. fetchSnapshotCore belongs to #2 and all code in storage manager is 3.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: fetchLatestSnapshotCore only know which type of snapshot needs to fetched as that creates url and based on that it can call different converters to convert into common format.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:210]: "snapshotDownloader" is just the stub which fetches the snapshot(initiates network call). It does not know what type of snapshot it is and what conversion needs to be done. If we want snapshotDownloader to do the conversion, then that conversion logic will be passed by "fetchLatestSnapshotCore" to snapshotDownloader as callback or maybe some arrgument telling what type of snapshot it is so that it can do the conversion as only "fetchLatestSnapshotCore " knows which type of snapshot is fetched.

**Comment by jatgarg** [comment]
Ping

File: packages/drivers/odsp-driver/src/createFile.ts [hunk 4]
```diff
@@ -81,14 +81,10 @@ export async function createNewFluidFile(
...
-        const snapshot: IOdspSnapshot = convertCreateNewSummaryTreeToIOdspSnapshot(createNewSummary, summaryHandle);
-        const protocolSummary = createNewSummary.tree[".protocol"] as ISummaryTree;
-        const documentAttributes = getDocAttributesFromProtocolSummary(protocolSummary);
-        const sequenceNumber = documentAttributes.sequenceNumber;
-
+        const snapshot: ISnapshotValue = convertCreateNewSummaryTreeToTreeAndBlobs(createNewSummary, summaryHandle);
         // caching the converted summary
-        const value: ISnapshotCacheValue = { snapshot, sequenceNumber };
-        await epochTracker.put(createCacheSnapshotKey(odspResolvedUrl), value);
+        const cacheValue: ISnapshotCacheValueWithVersion = { ...snapshot, version: 2 };
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/createFile.ts:86]: EpochTracker.put():
```typescript
    public async put(entry: IEntry, value: any) {
        assert(this._fluidEpoch !== undefined, 0x1dd /* "no epoch" */);
        const data: IVersionedValueWithEpoch = {
            value,
            version: persistedCacheValueVersion,
            fluidEpoch: this._fluidEpoch,
        };
```

Which means this code should not deal with versions.
Is this change intentional?

Also version of persistedCacheValueVersion should be bumped, as content takes different form, right?

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/createFile.ts:86]: "persistedCacheValueVersion" deals with cache data version and is for anything put in the cache. The SnapshotCacheValue Version represents version of snapshot data as not we can have different formats. So the host cache can contain cached snapshot and needs to be treated as such while this newer binary snapshot stored with version tells us what kind of snapshot it  is.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/createFile.ts:86]: ISnapshotCacheValueWithVersion just represents the value in   const data: IVersionedValueWithEpoch = {
            value,
            version: persistedCacheValueVersion,
            fluidEpoch: this._fluidEpoch,
        };

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:149]: That makes sense. But I think we need to chose between the two options:
1. snapshotDownloader is some basic wrapper around fetch() that does not really know type it returns. It should be marked as returning `unknown` and the caller should do appropriate casting based on its knowledge of what it works with.
2. we expect callback to return IOdspSnapshot, as that's what this function expects.

I think # 2 is more suiting here - it's a requirement of this function to have a callback that returns IOdspSnapshot. How it's implemented / provided (and does it simply cases unknown to IOdspSnapshot) - it does not really matter from POV of this function.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 3 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:211]: So if I got it right, we would need to rename fetchLatestSnapshotCore into something like fetchLatestSnapshotV1, add parallel fetchLatestSnapshotV2, and make fetchSnapshotWithRedeem() be the function that calls both versions depending on feature flag, right?

fetchLatestSnapshotCore is called twice from fetchSnapshotWithRedeem and fetchLatestSnapshotCore  has a lot of logic (resource timing telemetry) that would be useful for both paths, so I'd probably try to create fetchLatestSnapshotV1 & fetchLatestSnapshotV2 be responsible only for URI formatting and initial parsing / conversion of data, but keep fetchLatestSnapshotCore as a function on top of them (and the one where we do fork between two calls).

File: packages/drivers/odsp-driver/src/createFile.ts (hunk 4 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/createFile.ts:86]: I'd simplify and only support new format going forward. I.e. no matter what wire format is used, it is always converted to ISnapshotValue and that is what stored in cache and assumed to be in the cache. persistedCacheValueVersion is bumped to start ignoring old entries.
I think it's much simpler model then adding version on top of version and complicating both read & write code.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/createFile.ts:86]: Basically persistedCacheValueVersion's goal is precisely to deal with changes in caching schema format by ignoring old entries - there is some perf loss of not using data in old format, but it's very short lived, we should optimize toward simplicity...

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/createFile.ts:86]: For newer entries I am storing the ISnapshotValue only. I was adding version for figuring out old entries as I missed we can do that with persistedCacheValueVersion. So will remove extra version.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:211]: fetchSnapshotWithRedeem would just call to fetchLatestSnapshotCore. It is fetchLatestSnapshotCore which will decide what to fetch. I can add fetchLatestSnapshotV1 and V2 once I integrate the second format.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:149]: Requirement to get IOdspSnapshot if for older format. Once we integrate new format, then snapshotDownloader will not return IOdspSnapshot rather will return the newer format. So that is why I made snapshotDownloader generic and it does not need to know what type is being fetched.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:149]: Only the fetchLatestSnapshotCore will know what type is being fetched.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 1 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts]: Removed it.

File: packages/drivers/odsp-driver/src/contracts.ts [hunk 5]
```diff
@@ -202,7 +202,7 @@ export interface ICreateFileResponse {
 export interface IVersionedValueWithEpoch {
     value: any;
     fluidEpoch: string,
-    version: 2,
+    version: 3,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/contracts.ts:205]: I assume there is no way to use persistedCacheValueVersion here. Maybe add comment that it's same as persistedCacheValueVersion?

File: packages/drivers/odsp-driver/src/odspUtils.ts [hunk 6]
```diff
@@ -30,16 +31,18 @@ import {
...
 
 export const getWithRetryForTokenRefreshRepeat = "getWithRetryForTokenRefreshRepeat";
 
 /** Parse the given url and return the origin (host name) */
 export const getOrigin = (url: string) => new URL(url).origin;
 
-export interface ISnapshotCacheValue {
-    snapshot: IOdspSnapshot;
-    sequenceNumber: number | undefined;
+export interface ISnapshotValue {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspUtils.ts:41]: maybe ISnapshotContets?

File: packages/drivers/odsp-driver/src/odspSnapshotParser.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        odspSnapshot.blobs.forEach((blob) => {
+            assert(blob.encoding === "base64" || blob.encoding === undefined,
+                0x0a4 /* `Unexpected blob encoding type: '${blob.encoding}'` */);
+            blobsWithBufferContent.set(blob.id, stringToBuffer(blob.content, blob.encoding ?? "utf8"));
+        });
+    }
+    const val: ISnapshotValue = {
+        blobs: blobsWithBufferContent,
+        ops: odspSnapshot.ops ?? [],
+        sequenceNumber: odspSnapshot.trees && (odspSnapshot.trees[0] as any).sequenceNumber,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspSnapshotParser.ts:70]: It would be great to assert here that odspSnapshot.trees.length === 1.
I assume that relies on work Marcelo did recently to remove commits from response. If so, would be great to confirm it is enabled in prod and for all generations of files.
Also can you please rename IOdspSnapshot.trees to IOdspSnapshot.commits? I made it's type to be IOdspSnapshotCommit to better reflect meaning, but never renamed actual variables...

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (vladsud) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: I think we should remove template argument <T> here and use `unknown` return type - that much better reflects the intention here.

File: packages/drivers/odsp-driver/src/odspSnapshotParser.ts (hunk 7 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspSnapshotParser.ts:70]: Yes, I confirmed him before, he said by  mid-end of june. I will confirm once more.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: But we know the type, then why unknown?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 8]
```diff
@@ -364,32 +319,16 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
...
+        return patchedBlob;
     }
 
     public async readBlob(blobId: string): Promise<ArrayBufferLike> {
         const blob = await this.readBlobCore(blobId);
-        if (blob instanceof ArrayBuffer) {
-            return blob;
-        }
-        return stringToBuffer(blob.content, blob.encoding ?? "utf8");
+        return blob;
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:331]: you can simplify and just do
```typescript
return this.readBlobCore(blobId);
```

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 9]
```diff
@@ -516,18 +451,14 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
...
-            if (trees) {
-                this.initCommitCache(trees);
-                // versionId is the id of the first tree
-                if (trees.length > 0) {
-                    id = trees[0].id;
-                }
+            if (snapshotTree) {
+                id = snapshotTree.id;
+                assert(id !== undefined, "Root tree should contain the id");
+                this.initCommitCache(new Map([[id, snapshotTree]]));
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:461]: The name and typing no longer match. Also it always received exactly one entry.
So I'd change it to something like setRootTree(id, tree);

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 10]
```diff
@@ -705,34 +636,30 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
         }
     }
 
-    private async readCommit(id: string): Promise<IOdspSnapshotCommit | null> {
+    private async readCommit(id: string): Promise<api.ISnapshotTree | null> {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:639]: Similar feedback with naming, I think we expect to get snapshot ID, and name of function should not mention commit?

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: The callback provided does not know the type. It's only the caller who knows the type

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 10 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Changed it to readTree.

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: Yes, that is why I provided template as the caller would supply the type of response it expects. Isn't it fine? Sorry I don't understand if you are saying something else.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: It's not wrong, but it is sort of cheating, as for function to be truely generic, it needs to depend on T somewhere (i.e. one of the args should depend on T). Here, it's a hidden cast, I'd rather see it more direct in this function.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:148]: BTW, I started that cheating (in other places), but I think it's wrong :)

File: packages/drivers/odsp-driver/src/odspSnapshotParser.ts (hunk 7 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspSnapshotParser.ts]: Renamed trees to commits.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspSnapshotParser.ts]: I think we need to wait a bit more for adding that assert. With that change, we would also be able to remove some code where we used to handle that.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspSnapshotParser.ts]: Actually we cannot rename trees to commits. It is the type that we fetch from server and we get trees from server. Also I think with new format we are calling trees and IOdspSnapshot would be totally gone. So we can leave it as such.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspSnapshotParser.ts]: You are right about rename :)
If we are not asserting and getting two entries there, what does it mean? Don't we have .protocol commit in first entry, and simply ignore .app enrty?
